AngularJS Sample App

ireasor opened the pull request:
Here's a pull request for an AngularJS sample app.  I have added comments throughout; let me know if there is anything I should do to better clarify the way that this works.

ireasor committed:
Initial checkin for AngularJS integration sample.  Includes base structure and pages along with one component and moving from one page to another.

ireasor committed:
Finished building out the app and added some comments.

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/clientlibs/footerLib/.content.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jcr:root xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
+    jcr:primaryType="cq:ClientLibraryFolder"
+    categories="[apps.samples.angularjs-footerlib]"
```
can we namespace with `apps.acs-samples` to be crystal clear

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/clientlibs/footerLib/js.txt:
```diff
@@ -0,0 +1,13 @@
+angular/core/baseController.js
+angular/core/routingService.js
+
+angular/partialControllers/homePartialController.js
```
Is "partialController" naming an angularJS convention? Are these any different from normal controllers (dont these just inherit from a base controller)

Just seeing if we can reduce the number of points that have the potential to confuse (case in point; I am wondering why this isnt just named "controllers")

ireasor replied:
Partials are the piece of the page that is being loaded in dynamically
through AJAX via Angular routing.  The idea here was to group the
controllers that are being assigned by Angular routing together in one
place.  Unlike other controllers, you won't see these in the markup, only
in the routing configuration.  I can change it, but I didn't want readers
to be confused as to how these are assigned.  I do have a comment in the
top of each of those files, so maybe this is not necessary.

On Sun, Mar 8, 2015 at 8:41 AM, david g. notifications@github.com wrote:

> In
> content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/clientlibs/footerLib/js.txt
> https://github.com/Adobe-Consulting-Services/acs-aem-samples/pull/26#discussion_r26006152
> :
> 
> > @@ -0,0 +1,13 @@
> > +#base=js
> > +
> > +angular/core/main.js
> > +angular/core/baseController.js
> > +angular/core/routingService.js
> > +
> > +angular/partialControllers/homePartialController.js
> 
> Is "partialController" naming an angularJS convention? Are these any
> different from normal controllers (dont these just inherit from a base
> controller)
> 
> Just seeing if we can reduce the number of points that have the potential
> to confuse (case in point; I am wondering why this isnt just named
> "controllers")
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Adobe-Consulting-Services/acs-aem-samples/pull/26/files#r26006152
> .

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/clientlibs/footerLib/js/angular/core/main.js:
```diff
@@ -0,0 +1,30 @@
+/*
+ * The routeProvider configuration maps URL paths to AJAX content to be loaded into the ng-view container defined in the containerPage component.
+ * Note that the templateUrls are each using a .content Sling selector.  This allows us to switch out the content portion of the containerPage with that of the partials.
+ */
```
I think its worth noting that the `.content` selector is a custom abstraction defined by the implementation of the angularjs-integration base-page, and not some "magical" AEM selector.

ireasor replied:
This is 100% sling.  Angular has no concept of selectors.  I updated the
comment to make this more clear.  Let me know if it is still confusing.

On Sun, Mar 8, 2015 at 8:45 AM, david g. notifications@github.com wrote:

> In
> content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/clientlibs/footerLib/js/angular/core/main.js
> https://github.com/Adobe-Consulting-Services/acs-aem-samples/pull/26#discussion_r26006199
> :
> 
> > @@ -0,0 +1,30 @@
> > +/_global angular_/
> > +
> > +/*
> > - \* This is the definition of the application module itself as well as core configurations like routing and the http service.
> > - \* Since this file defines that module that all of our controllers and services are part of, it needs to be listed first in the clientlibs js.txt file.
> > - _/
> >   +var sampleModule = angular.module('angularjs-sample', ['ngRoute']);
> >   +
> >   +/_
> > - \* The routeProvider configuration maps URL paths to AJAX content to be loaded into the ng-view container defined in the containerPage component.
> > - \* Note that the templateUrls are each using a .content Sling selector.  This allows us to switch out the content portion of the containerPage with that of the partials.
> > - */
> 
> I think its worth noting that the .content selector is a custom
> abstraction defined by the implementation of the angularjs-integration
> base-page, and not some "magical" AEM selector.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Adobe-Consulting-Services/acs-aem-samples/pull/26/files#r26006199
> .

davidjgonzalez replied:
Yup - i get that; just want it to be clear to less familiar folks reading through :) .. If i was new to this i would think "what is this magic .content selector being used, why is it being used, can i use a different one, how do i implement it, etc." .. the simple answer is its a selector that binds to a content.jsp script on the page that is responsible for rendering only the markup necessary for that view and not any HTML header/footer data. 

Thats easy for you and I to know that, but if youve never implemented a single page angular app in AEM its not totally straightforward.

ireasor replied:
Yup.  I have updated the comment to be more clear.  I'm just validating
everything before I push again.

On Sun, Mar 8, 2015 at 11:41 AM, david g. notifications@github.com wrote:

> In
> content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/clientlibs/footerLib/js/angular/core/main.js
> https://github.com/Adobe-Consulting-Services/acs-aem-samples/pull/26#discussion_r26007636
> :
> 
> > @@ -0,0 +1,30 @@
> > +/_global angular_/
> > +
> > +/*
> > - \* This is the definition of the application module itself as well as core configurations like routing and the http service.
> > - \* Since this file defines that module that all of our controllers and services are part of, it needs to be listed first in the clientlibs js.txt file.
> > - _/
> >   +var sampleModule = angular.module('angularjs-sample', ['ngRoute']);
> >   +
> >   +/_
> > - \* The routeProvider configuration maps URL paths to AJAX content to be loaded into the ng-view container defined in the containerPage component.
> > - \* Note that the templateUrls are each using a .content Sling selector.  This allows us to switch out the content portion of the containerPage with that of the partials.
> > - */
> 
> Yup - i get that; just want it to be clear to less familiar folks reading
> through :) .. If i was new to this i would think "what is this magic
> .content selector being used, why is it being used, can i use a different
> one, how do i implement it, etc." .. the simple answer is its a selector
> that binds to a content.jsp script on the page that is responsible for
> rendering only the markup necessary for that view and not any HTML
> header/footer data.
> 
> Thats easy for you and I to know that, but if youve never implemented a
> single page angular app in AEM its not totally straightforward.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Adobe-Consulting-Services/acs-aem-samples/pull/26/files#r26007636
> .

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/clientlibs/footerLib/js/angular/core/main.js:
```diff
@@ -0,0 +1,30 @@
+sampleModule.config(['$routeProvider', function ($routeProvider) {
+        $routeProvider.
+                when('/home', {
+                    templateUrl: '/content/acs-samples/angularjs-integration/home.content.html',
```
Is it best practice to hardcode these URIs? Is there a simple, clean way around it? 

At the same time dont want to add too much complexity and orchestration that would distract from the fundamentals of the integration.

ireasor replied:
That's a good question.  I guess you could build out some tool to let an
author configure these and load them in via a call back to the repo or an
OSGi service.  It would probably add a fair amount of complexity here,
though.  In our actual implementation, we took this approach since there
was a predefined booking funnel that was not likely to change without the
need for developer intervention anyway.

On Sun, Mar 8, 2015 at 8:46 AM, david g. notifications@github.com wrote:

> In
> content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/clientlibs/footerLib/js/angular/core/main.js
> https://github.com/Adobe-Consulting-Services/acs-aem-samples/pull/26#discussion_r26006206
> :
> 
> > +/_global angular_/
> > +
> > +/*
> > - \* This is the definition of the application module itself as well as core configurations like routing and the http service.
> > - \* Since this file defines that module that all of our controllers and services are part of, it needs to be listed first in the clientlibs js.txt file.
> > - _/
> >   +var sampleModule = angular.module('angularjs-sample', ['ngRoute']);
> >   +
> >   +/_
> > - \* The routeProvider configuration maps URL paths to AJAX content to be loaded into the ng-view container defined in the containerPage component.
> > - \* Note that the templateUrls are each using a .content Sling selector.  This allows us to switch out the content portion of the containerPage with that of the partials.
> > - */
> >   +sampleModule.config(['$routeProvider', function ($routeProvider) {
> > -        $routeProvider.
> > -                when('/home', {
> > -                    templateUrl: '/content/acs-samples/angularjs-integration/home.content.html',
> 
> Is it best practice to hardcode these URIs? Is there a simple, clean way
> around it?
> 
> At the same time dont want to add too much complexity and orchestration
> that would distract from the fundamentals of the integration.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Adobe-Consulting-Services/acs-aem-samples/pull/26/files#r26006206
> .

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/clientlibs/footerLib/js/angular/core/main.js:
```diff
@@ -0,0 +1,30 @@
+ */
+sampleModule.config(['$routeProvider', function ($routeProvider) {
+        $routeProvider.
+                when('/home', {
```
May be worth noting w a one-liner that the routes are NOT the requested URIs;

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/clientlibs/headerLib/.content.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jcr:root xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
+    jcr:primaryType="cq:ClientLibraryFolder"
+    categories="[apps.samples.angularjs-headerlib]"/>
```
`categories="[apps.acs-samples.angularjs-headerlib]"/>`

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/components/checkoutForm/.content.xml:
```diff
@@ -0,0 +1,7 @@
+    jcr:primaryType="cq:Component"
+    jcr:title="Checkout Form"
+    allowedParents="[*/parsys]"
+    componentGroup="AngularJS - Sample"/>
```
`ACS Samples - AngularJS App`

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/components/checkoutForm/dialog.xml:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jcr:root xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
+    jcr:primaryType="cq:Dialog"
+    title="Checkout form"
```
Checkout Form

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-samples/angularjs-integration/templates/pagePartial/.content.xml:
```diff
@@ -0,0 +1,10 @@
+    jcr:description="The template for the AngularJS integration page partials"
+    jcr:primaryType="cq:Template"
+    jcr:title="AngularJS Integration Page Partial Template"
+    allowedPaths="[/content(.*)]">
```
/content/acs-samples/angular-app(.*)

davidjgonzalez commented on content/src/main/content/jcr_root/content/.content.xml:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
What is this file for? We shouldnt need this.

davidjgonzalez commented on content/src/main/content/jcr_root/content/acs-samples/.content.xml:
```diff
@@ -0,0 +1,15 @@
+<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0" xmlns:nt="http://www.jcp.org/jcr/nt/1.0"
+    jcr:primaryType="cq:Page">
+    <jcr:content
+        cq:lastModified="{Date}2014-04-22T10:11:24.002+01:00"
```
junk properties

davidjgonzalez commented on content/src/main/content/jcr_root/content/acs-samples/angularjs-integration/.content.xml:
```diff
@@ -0,0 +1,14 @@
+<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
+    jcr:primaryType="cq:Page">
+    <jcr:content
+        cq:lastModified="{Date}2015-02-27T11:18:53.196-07:00"
```
Junk properties

davidjgonzalez commented on content/src/main/content/jcr_root/content/acs-samples/angularjs-integration/checkout/.content.xml:
```diff
@@ -0,0 +1,23 @@
+<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0" xmlns:nt="http://www.jcp.org/jcr/nt/1.0"
+    jcr:primaryType="cq:Page">
+    <jcr:content
+        cq:lastModified="{Date}2015-03-02T11:05:41.203-07:00"
```
Junk properties

davidjgonzalez commented on content/src/main/content/jcr_root/content/acs-samples/angularjs-integration/home/.content.xml:
```diff
@@ -0,0 +1,24 @@
+<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0" xmlns:nt="http://www.jcp.org/jcr/nt/1.0"
+    jcr:primaryType="cq:Page">
+    <jcr:content
+        cq:lastModified="{Date}2015-02-27T15:00:05.655-07:00"
```
junk props

davidjgonzalez commented on content/src/main/content/jcr_root/content/acs-samples/angularjs-integration/home/.content.xml:
```diff
@@ -0,0 +1,24 @@
+            jcr:primaryType="nt:unstructured"
+            sling:resourceType="foundation/components/parsys">
+            <searchbox
+                jcr:created="{Date}2015-02-27T14:59:44.909-07:00"
```
junk props

davidjgonzalez commented:
@ireasor bunch of minor stuff; We should change the node naming and property naming to adhere w jackrabbit conventions; We've been moving forward with

node names are: all lower case, dashes for spaces (this would include jsps as they eventually manifest as nodes in JCR)
property names: camelCase

I can do this after the merge though as its rather annoying.

davidjgonzalez commented:
Also - WDYT about calling this "AngularJS App" rather than "integration" ? ..  or "AngularJS One-Page App" .. ?

ireasor committed:
Implemented code review suggestions.
Also changed the binding strategy to bind to values on the data object rather than the values themselves.
When reassigning a value, it is easy to lose reference to the object in the old method.  By binding to the data object, you can see an update from the binding even when reassigning the property.
Fixed a bug in the searchEndpoint where I should have been checking for > -1 rather than > 0.

ireasor commented:
I went through and implemented your suggestions as well as improving the way I was setting up binding and fixed a bug in the endpoint.

davidjgonzalez commented:
@ireasor Awesome! Thanks a million for this - i think folks will find it quite helpful.
Will merge and add to the Samples Index page!
